// @babylonjs/core/Particles/Node/Blocks/Update/updateRemapBlock.js
import { Vector2, Vector4 } from "../../../../Maths/math.vector.js";
import { RegisterClass } from "../../../../Misc/typeStore.js";
import { NodeParticleBlock } from "../../nodeParticleBlock.js";
import { NodeParticleBlockConnectionPointTypes } from "../../Enums/nodeParticleBlockConnectionPointTypes.js";
import { _ConnectAtTheEnd } from "../../../Queue/executionQueue.js";
/**
 * Block used to update the remap value of a particle
 */
export class UpdateRemapBlock extends NodeParticleBlock {
    /**
     * Create a new UpdateRemapBlock
     * @param name defines the block name
     */
    constructor(name) {
        super(name);
        this.registerInput("particle", NodeParticleBlockConnectionPointTypes.Particle);
        this.registerInput("remapColor", NodeParticleBlockConnectionPointTypes.Vector2, true);
        this.registerInput("remapAlpha", NodeParticleBlockConnectionPointTypes.Vector2, true);
        this.registerOutput("output", NodeParticleBlockConnectionPointTypes.Particle);
    }
    /**
     * Gets the particle component
     */
    get particle() {
        return this._inputs[0];
    }
    /**
     * Gets the remapColor input component
     */
    get remapColor() {
        return this._inputs[1];
    }
    /**
     * Gets the remapAlpha input component
     */
    get remapAlpha() {
        return this._inputs[2];
    }
    /**
     * Gets the output component
     */
    get output() {
        return this._outputs[0];
    }
    /**
     * Gets the current class name
     * @returns the class name
     */
    getClassName() {
        return "UpdateRemapBlock";
    }
    /**
     * Builds the block
     * @param state defines the current build state
     */
    _build(state) {
        const system = this.particle.getConnectedValue(state);
        this.output._storedValue = system;
        if (!this.remapColor.isConnected && !this.remapAlpha.isConnected) {
            return;
        }
        const processRemap = (particle) => {
            state.particleContext = particle;
            state.systemContext = system;
            const colorRemap = this.remapColor.isConnected ? this.remapColor.getConnectedValue(state) : new Vector2(0, 1);
            const alphaRemap = this.remapAlpha.isConnected ? this.remapAlpha.getConnectedValue(state) : new Vector2(0, 1);
            if (!particle.remapData) {
                particle.remapData = new Vector4(0, 1, 0, 1);
            }
            if (colorRemap) {
                particle.remapData.x = colorRemap.x;
                particle.remapData.y = colorRemap.y;
            }
            if (alphaRemap) {
                particle.remapData.z = alphaRemap.x;
                particle.remapData.w = alphaRemap.y;
            }
        };
        const remapProcessing = {
            process: processRemap,
            previousItem: null,
            nextItem: null,
        };
        if (system._updateQueueStart) {
            _ConnectAtTheEnd(remapProcessing, system._updateQueueStart);
        }
        else {
            system._updateQueueStart = remapProcessing;
        }
    }
}
RegisterClass("BABYLON.UpdateRemapBlock", UpdateRemapBlock);
